google/cloud/forseti/scanner/audit/retention_rules_engine.py

```diff
@@ -0,0 +1,327 @@
+
+    def build_rule_book(self, global_configs=None):
+        """Build RetentionRuleBook from the rules definition file.
+
+        Args:
+            global_configs (dict): Global configurations.
+        """
+        self.rule_book = RetentionRuleBook(self._load_rule_definitions())
+
+    def find_buckets_violations(self, buckets_lifecycle, force_rebuild=False):
+        """Determine whether bucket lifecycle violates rules.
+
+        Args:
+            buckets_lifecycle (retention_bucket.RetentionBucket): Object
+                containing lifecycle data
+            force_rebuild (bool): If True, rebuilds the rule book. This will
+                reload the rules definition file and add the rules to the book.
+
+        Returns:
+             generator: A generator of rule violations.
+        """
+        if self.rule_book is None or force_rebuild:
+            self.build_rule_book()
+
+        violations = itertools.chain()
+
+        resource_rules = self.rule_book.get_resource_rules(_APPLY_TO_BUCKETS)
+
+        resource_ancestors = (relationship.find_ancestors(
+            buckets_lifecycle, buckets_lifecycle.full_name))
+        for related_resources in resource_ancestors:
+            rules = resource_rules[related_resources]
+            for rule in rules:
+                violations = itertools.chain(
+                    violations,
+                    rule.find_buckets_violations(buckets_lifecycle))
+
+        return set(violations)
+
+
+def get_bucket_retention_range(rule_def, rule_index):
+    """Add a rule to the rule book.
+
+    Args:
+        rule_def (dict): A dictionary containing rule definition
+            properties.
+        rule_index (int): The index of the rule from the rule definitions.
+            Assigned automatically when the rule book is built.
+
+    Returns:
+        pair: the minimum and maximum value of the Age.
+    """
+    minimum_retention = rule_def.get('minimum_retention', None)
+    maximum_retention = rule_def.get('maximum_retention', None)
+    if minimum_retention is None and maximum_retention is None:
+        raise audit_errors.InvalidRulesSchemaError(
+            'Lack of minimum_retention and '
+            'maximum_retention in rule {}'.format(rule_index))
+    elif minimum_retention != None and maximum_retention != None:
+        if minimum_retention > maximum_retention:
+            raise audit_errors.InvalidRulesSchemaError(
+                'minimum_retention larger than '
+                'maximum_retention in rule {}'.format(rule_index))
+    return (minimum_retention, maximum_retention)
+
+
+class RetentionRuleBook(bre.BaseRuleBook):
+    """The RuleBook for Retention resources."""
+
+    def __init__(self, rule_defs=None):
+        """Initialization.
+
+        Args:
+            rule_defs (dict): rule definitons
+        """
+        super(RetentionRuleBook, self).__init__()
+        self._rules_sema = threading.BoundedSemaphore(value=1)
+
+        self.resource_rules_map = {
```

Comment:
so this is a dict of dicts? do we need that? would just a dict of resource type to list of rules work?

Reply:
I believe we need not only the type but also the id to identify the rules so the book should be dict[type][id] something like that.

Reply:
Since many other scanners have the same name with the same purpose, I want to keep the consistency.